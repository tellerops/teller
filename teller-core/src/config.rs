use std::cmp::Ordering;
use std::path::PathBuf;
use std::{
    collections::{BTreeMap, HashMap},
    path::Path,
};

use fs_err as fs;
use serde_derive::{Deserialize, Serialize};
use teller_providers::config::{PathMap, ProviderCfg, KV};
use teller_providers::providers::ProviderKind;
use tera::{Context, Tera};

use crate::Result;

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct Config {
    pub providers: BTreeMap<String, ProviderCfg>,
}

#[derive(Serialize)]
pub struct RenderTemplate {
    pub providers: Vec<ProviderKind>,
}

fn apply_eqeq(config: &mut Config) {
    config.providers.iter_mut().for_each(|(_name, provider)| {
        provider.maps.iter_mut().for_each(|pm| {
            pm.keys.iter_mut().for_each(|(k, v)| {
                // THINK: replace with:
                // 1. templating: {{id}} (identity), {{snake_case}} (snake case it)
                // 2. other symbols: == id, ^^ capitalize, snake case __ lower snake case
                if v == "==" {
                    v.clone_from(k);
                }
            });
        });
    });
}

impl Config {
    /// Config from text
    ///
    /// # Errors
    ///
    /// This function will return an error if serialization fails
    pub fn with_vars(text: &str, vars: &HashMap<String, String>) -> Result<Self> {
        let rendered_text = Tera::one_off(text, &Context::from_serialize(vars)?, false)?;
        let mut config: Self = serde_yaml::from_str(&rendered_text)?;

        apply_eqeq(&mut config);

        Ok(config)
    }

    /// Config from text
    ///
    /// # Errors
    ///
    /// This function will return an error if serialization fails
    pub fn from_text(text: &str) -> Result<Self> {
        Self::with_vars(text, &HashMap::new())
    }

    /// Config from file
    ///
    /// # Errors
    ///
    /// This function will return an error if IO fails
    pub fn from_path(path: &Path) -> Result<Self> {
        Self::from_text(&fs::read_to_string(path)?)
    }

    /// Create configuration template file
    ///
    /// # Errors
    /// When could not convert config to string
    pub fn render_template(data: &RenderTemplate) -> Result<String> {
        let res: BTreeMap<String, ProviderCfg> = data
            .providers
            .iter()
            .map(|p| {
                (
                    format!("{p}_1"),
                    ProviderCfg {
                        kind: p.clone(),
                        maps: vec![PathMap::from_path("example/dev")],
                        ..ProviderCfg::default()
                    },
                )
            })
            .collect();

        let config = Self { providers: res };

        let a: String = serde_yaml::to_string(&config)?;
        Ok(a)
    }
}

#[derive(Debug, Clone, Serialize, Eq, PartialEq)]
pub struct Match {
    pub path: PathBuf,
    pub position: Option<(usize, usize)>,
    pub offset: usize,
    pub query: KV,
}

impl PartialOrd for Match {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Match {
    fn cmp(&self, other: &Self) -> Ordering {
        let query_cmp = self.query.cmp(&other.query);

        if query_cmp != Ordering::Equal {
            return query_cmp;
        }

        self.offset.cmp(&other.offset)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;
    #[test]
    fn load_config() {
        std::env::set_var("TEST_LOAD_1", "DEV");
        let config = Config::from_path(Path::new("fixtures/config.yml")).unwrap();
        assert_eq!(config.providers.len(), 2);
        assert_yaml_snapshot!(config);
    }

    #[test]
    fn can_render_template_config() {
        let data = RenderTemplate {
            providers: vec![ProviderKind::Inmem, ProviderKind::Dotenv],
        };

        let config = Config::render_template(&data).unwrap();
        assert_yaml_snapshot!(config);
    }
}
