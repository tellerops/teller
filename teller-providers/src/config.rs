use std::cmp::Ordering;
use std::collections::BTreeMap;

use serde_derive::{Deserialize, Serialize};

use crate::providers::ProviderKind;

fn is_default<T: Default + PartialEq>(t: &T) -> bool {
    t == &T::default()
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct ProviderCfg {
    #[serde(rename = "kind")]
    pub kind: ProviderKind,
    #[serde(rename = "options", skip_serializing_if = "Option::is_none")]
    pub options: Option<serde_json::Value>,
    #[serde(rename = "name", skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub maps: Vec<PathMap>,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub enum Sensitivity {
    #[default]
    None,
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct ProviderInfo {
    pub kind: ProviderKind,
    pub name: String,
}
#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct PathInfo {
    pub id: String,
    pub path: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct MetaInfo {
    pub sensitivity: Sensitivity,
    pub redact_with: Option<String>,
    pub source: Option<String>,
    pub sink: Option<String>,
}
#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct KV {
    pub value: String,
    pub key: String, // mapped-to key
    pub from_key: String,
    pub path: Option<PathInfo>, // always toplevel
    pub provider: Option<ProviderInfo>,
    pub meta: Option<MetaInfo>,
}

impl PartialOrd for KV {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for KV {
    fn cmp(&self, other: &Self) -> Ordering {
        if let (Some(provider), Some(other_provider)) = (&self.provider, &other.provider) {
            let provider_cmp = provider.kind.cmp(&other_provider.kind);
            if provider_cmp != Ordering::Equal {
                return provider_cmp;
            }
        }

        self.key.cmp(&other.key)
    }
}

impl KV {
    #[must_use]
    pub fn to_data(kvs: &[Self]) -> BTreeMap<String, String> {
        let mut data = BTreeMap::new();
        for kv in kvs {
            data.insert(kv.key.clone(), kv.value.clone());
        }
        data
    }

    #[must_use]
    pub fn from_data(
        data: &BTreeMap<String, String>,
        pm: &PathMap,
        provider: &ProviderInfo,
    ) -> Vec<Self> {
        // map all of the data found
        if pm.keys.is_empty() {
            data.iter()
                .map(|(k, v)| Self::from_value(v, k, k, pm, provider.clone()))
                .collect::<Vec<_>>()
        } else {
            // selectively map only keys from pathmap
            pm.keys
                .iter()
                .filter_map(|(from_key, to_key)| {
                    data.get(from_key).map(|found_val| {
                        Self::from_value(found_val, from_key, to_key, pm, provider.clone())
                    })
                })
                .collect::<Vec<_>>()
        }
    }
    #[must_use]
    pub fn from_value(
        found_val: &str,
        from_key: &str,
        to_key: &str,
        pm: &PathMap,
        provider: ProviderInfo,
    ) -> Self {
        Self {
            value: found_val.to_string(),
            key: to_key.to_string(),
            from_key: from_key.to_string(),
            path: Some(PathInfo {
                path: pm.path.clone(),
                id: pm.id.to_string(),
            }),
            provider: Some(provider),
            meta: Some(MetaInfo {
                sensitivity: pm.sensitivity.clone(),
                redact_with: pm.redact_with.clone(),
                source: pm.source.clone(),
                sink: pm.sink.clone(),
            }),
        }
    }
    #[must_use]
    pub fn from_literal(path: &str, key: &str, value: &str, provider: ProviderInfo) -> Self {
        Self {
            value: value.to_string(),
            key: key.to_string(),
            from_key: key.to_string(),
            path: Some(PathInfo {
                id: path.to_string(),
                path: path.to_string(),
            }),
            provider: Some(provider),
            ..Default::default()
        }
    }

    /// represents a KV without any source (e.g. created manually by a user, pending insert to
    /// one of the providers)
    #[must_use]
    pub fn from_kv(key: &str, value: &str) -> Self {
        Self {
            value: value.to_string(),
            key: key.to_string(),
            from_key: key.to_string(),
            ..Default::default()
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct PathMap {
    pub id: String,
    #[serde(rename = "protocol", skip_serializing_if = "Option::is_none")]
    pub protocol: Option<String>,
    #[serde(rename = "path")]
    pub path: String,
    #[serde(default, rename = "keys", skip_serializing_if = "is_default")]
    pub keys: BTreeMap<String, String>,
    #[serde(default, rename = "decrypt", skip_serializing_if = "is_default")]
    pub decrypt: bool,
    #[serde(default, rename = "sensitivity", skip_serializing_if = "is_default")]
    pub sensitivity: Sensitivity,
    #[serde(
        default,
        rename = "redact_with",
        skip_serializing_if = "Option::is_none"
    )]
    pub redact_with: Option<String>,
    #[serde(default, rename = "source", skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
    #[serde(default, rename = "sink", skip_serializing_if = "Option::is_none")]
    pub sink: Option<String>,
    // ignore population if optional + we got error
    #[serde(default, rename = "optional", skip_serializing_if = "is_default")]
    pub optional: bool,
}

impl PathMap {
    #[must_use]
    pub fn from_path(path: &str) -> Self {
        Self {
            path: path.to_string(),
            ..Default::default()
        }
    }
}
