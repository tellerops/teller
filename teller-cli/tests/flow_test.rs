use std::collections::HashMap;

use dockertest_server::{
    servers::{
        cloud::{LocalStackServer, LocalStackServerConfig},
        hashi::{VaultServer, VaultServerConfig},
    },
    Test,
};
use fs_err as fs;
use insta::assert_debug_snapshot;
use teller_core::config::Config;
use teller_providers::{
    config::{ProviderInfo, KV},
    providers::ProviderKind,
    registry::Registry,
};

fn build_providers() -> Test {
    let mut test = Test::new();
    test.register(
        VaultServerConfig::builder()
            .port(9200)
            .version("1.8.2".into())
            .build()
            .unwrap(),
    );
    test.register(
        LocalStackServerConfig::builder()
            .env(
                vec![(
                    "SERVICES".to_string(),
                    "iam,sts,ssm,kms,secretsmanager".to_string(),
                )]
                .into_iter()
                .collect(),
            )
            .port(4561)
            .version("2.0.2".into())
            .build()
            .unwrap(),
    );
    test
}

#[test]
#[cfg(not(windows))]
fn providers_smoke_test() {
    use std::{env, time::Duration};

    if env::var("RUNNER_OS").unwrap_or_default() == "macOS" {
        return;
    }

    let test = build_providers();

    test.run(|instance| async move {
        let user = "linus";
        let password = "torvalds123";
        let vault_server: VaultServer = instance.server();
        // banner is not enough for vault, we have to wait for the image to stabilize
        tokio::time::sleep(Duration::from_secs(2)).await;

        let localstack_server: LocalStackServer = instance.server();

        fs::write(
            "fixtures/flow.env",
            r"
FOO=bar
HELLO=world
",
        )
        .unwrap();
        let mut vars = HashMap::new();
        vars.insert("address".to_string(), vault_server.external_url());
        vars.insert("token".to_string(), vault_server.token);
        vars.insert("endpoint_url".to_string(), localstack_server.external_url());

        let config = Config::with_vars(
            &fs::read_to_string("fixtures/flow_test.yml").unwrap(),
            &vars,
        )
        .unwrap();
        let registry = Registry::new(&config.providers).await.unwrap();

        // (1) start: put on hashi
        let hashi = registry.get("hashi_1").unwrap();
        let hashi_pm0 = &config.providers.get("hashi_1").unwrap().maps[0];
        hashi
            .put(
                hashi_pm0,
                &[
                    KV::from_value(
                        user,
                        "USER",
                        "USER",
                        hashi_pm0,
                        ProviderInfo {
                            kind: ProviderKind::Hashicorp,
                            name: "test".to_string(),
                        },
                    ),
                    KV::from_value(
                        password,
                        "PASSWORD",
                        "PASSWORD",
                        hashi_pm0,
                        ProviderInfo {
                            kind: ProviderKind::Hashicorp,
                            name: "test".to_string(),
                        },
                    ),
                ],
            )
            .await
            .unwrap();
        let res = hashi.get(hashi_pm0).await;
        assert_debug_snapshot!("flow-test-hashi-0", res);

        // (2) push results into secretsmanager
        let kvs = res.unwrap();

        let smgr = registry.get("sm_1").unwrap();
        let smgr_pm0 = &config.providers.get("sm_1").unwrap().maps[0];
        smgr.put(smgr_pm0, &kvs[..]).await.unwrap();
        let res = smgr.get(smgr_pm0).await;
        assert_debug_snapshot!("flow-test-smgr-0", res);

        // (3) push results into ssm - remember it has custom key mapping so
        // check that in snapshots (USER -> USER_NAME, and drops the pass)
        let kvs = res.unwrap();

        let ssm = registry.get("ssm_1").unwrap();
        let ssm_pm0 = &config.providers.get("ssm_1").unwrap().maps[0];
        ssm.put(ssm_pm0, &kvs[..]).await.unwrap();
        let res = ssm.get(ssm_pm0).await;
        assert_debug_snapshot!("flow-test-ssm-0", res);

        // (4) lastly, write it into dotenv file, read it back, and snapshot it
        let kvs = res.unwrap();

        let dot = registry.get("dot_1").unwrap();
        let dot_pm0 = &config.providers.get("dot_1").unwrap().maps[0];
        dot.put(dot_pm0, &kvs[..]).await.unwrap();
        let res = dot.get(dot_pm0).await;
        assert_debug_snapshot!("flow-test-dot-0", res);
    });
}
