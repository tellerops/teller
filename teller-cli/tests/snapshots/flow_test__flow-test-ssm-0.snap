---
source: teller-cli/tests/flow_test.rs
expression: res
---
Ok(
    [
        KV {
            value: "linus",
            key: "USER_NAME",
            from_key: "USER",
            path: Some(
                PathInfo {
                    id: "pg-dev",
                    path: "/dev/postgres",
                },
            ),
            provider: Some(
                ProviderInfo {
                    kind: SSM,
                    name: "ssm_1",
                },
            ),
            meta: Some(
                MetaInfo {
                    sensitivity: None,
                    redact_with: None,
                    source: None,
                    sink: None,
                },
            ),
        },
    ],
)
